test/es-module/test-vm-main-context-default-loader.js

```diff
@@ -0,0 +1,170 @@
+    result = await imported;
+    assert.strictEqual(result.constants.F_OK, fs.constants.F_OK);
+  }
+
+  const modulePath = fixtures.path('es-modules', 'message.mjs');
+  fs.copyFileSync(modulePath, tmpdir.resolve('message.mjs'));
+
+  {
+    const namespace = await import(modulePath);
+    const script = new Script('import("./message.mjs")', options);
+    const result = await script.runInThisContext();
+    assert.deepStrictEqual(result, namespace);
+  }
+
+  {
+    const namespace = await import(modulePath);
+    const imported = compileFunction('return import("./message.mjs")', [], options)();
+    const result = await imported;
+    assert.deepStrictEqual(result, namespace);
+  }
+}
+
+async function main() {
+  {
+    // Importing with absolute path as filename.
+    tmpdir.refresh();
+    // Note that we must use different file names for different tests otherwise
+    // we would get cached results.
+    const filename = tmpdir.resolve('index1.js');
+    const options = {
+      filename,
+      importModuleDynamically: USE_MAIN_CONTEXT_DEFAULT_LOADER
+    };
+    await testNotFoundErrors(options);
+    await testLoader(options);
+  }
+
+  {
+    // Importing with file:// URL as filename.
+    tmpdir.refresh();
+    // Note that we must use different file names for different tests otherwise
+    // we would get cached results.
+    const filename = url.pathToFileURL(tmpdir.resolve('index2.js')).href;
+    const options = {
+      filename,
+      importModuleDynamically: USE_MAIN_CONTEXT_DEFAULT_LOADER
+    };
+    await testNotFoundErrors(options);
+    await testLoader(options);
+  }
+
+  {
+    // For undefined or non-path/URL filenames, import() should resolve to the cwd.
+    tmpdir.refresh();
+    process.chdir(tmpdir.path);
+    const undefinedOptions = {
+      importModuleDynamically: USE_MAIN_CONTEXT_DEFAULT_LOADER
+    };
+    const nonPathOptions = {
+      filename: 'non-path',
+      importModuleDynamically: USE_MAIN_CONTEXT_DEFAULT_LOADER
+    };
+    // Run the error tests first to avoid caching.
+    await testNotFoundErrors(undefinedOptions);
+    await testNotFoundErrors(nonPathOptions);
+
+    // createContext() with null referrer also resolves to cwd.
+    {
+      const options = {
+        importModuleDynamically: USE_MAIN_CONTEXT_DEFAULT_LOADER,
+      };
+      const ctx = createContext({}, options);
+      const s = new Script('Promise.resolve("import(\'./message.mjs\')").then(eval)', {
+        importModuleDynamically: common.mustNotCall(),
+      });
+      // Use try-catch for better async stack traces in the logs.
+      let result = 'uncaught-fallback';
+      try {
+        result = await s.runInContext(ctx);
+      } catch (error) {
+        assert.strictEqual(error.code, 'ERR_MODULE_NOT_FOUND');
+      } finally {
+        assert.strictEqual(
+          result, 'uncaught-fallback',
+          `import from vm.createContext() with ${inspect(options)} should throw ERR_MODULE_NOT_FOUND`);
+      }
```

Comment:
What do you mean "better async traces"? If `assert.rejects` generates worse async traces, shouldn't we patch it on a separate PR? The code is arguably much more readable than the try/catch pattern IMO
```suggestion
      await assert.rejects(s.runInContext(ctx), { code: 'ERR_MODULE_NOT_FOUND' });
```

Reply:
I think during local testing some frames disappeared when I used `assert.rejects`, but for the code being replaced here now it looks equivalent. Whether `assert.rejects` should be patched is probably out of scope in this PR.